api/traces/v1/trace_rbac.go

```diff
@@ -0,0 +1,227 @@
+package v1
+
+import (
+	"bytes"
+	"compress/flate"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+
+	// nolint:staticcheck
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/grafana/tempo/pkg/tempopb"
+	commonv1 "github.com/grafana/tempo/pkg/tempopb/common/v1"
+	tracev1 "github.com/grafana/tempo/pkg/tempopb/trace/v1"
+
+	apilogsv1 "github.com/observatorium/api/api/logs/v1"
+)
+
+const (
+	namespaceAttributeKey = "k8s.namespace.name"
+	serviceAttributeKey   = "service.name"
+)
+
+func WithTraceQLNamespaceSelect() func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if !strings.Contains(r.URL.Path, "/api/search") {
+				next.ServeHTTP(w, r)
+				return
+			}
+
+			query := r.URL.Query()
+			q := query.Get("q")
+			traceQL, err := url.QueryUnescape(q)
+			if err != nil {
+				next.ServeHTTP(w, r)
+				return
+			}
+			if traceQL == "" {
+				traceQL = "{ }"
+			}
+			traceQL = traceQL + " | select(resource.k8s.namespace.name)"
+			query.Set("q", traceQL)
+			r.URL.RawQuery = query.Encode()
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+func responseRBACModifier(log log.Logger) func(response *http.Response) error {
```

Comment:
General curiosity question, I'm certainly missing context but AFAIU, in the current implementation we are using OPA to extract the namespaces and then we are doing the authorization bit on the response in obs/api instead of OPA doing authorization on the query itself. Why did we go with such an approach?

Reply:
This predates my involvement in the project. I think they wanted to separate that functionality into an external component.